### PR TITLE
Api user management

### PR DESCRIPTION
**Pull request by swietlikm**
- added logout view
- added password_change view
- added user details view (get/patch)
- added tests for new views
- updated CustomUser model

### CODE REVIEW

File: Dshop/apps/users/api_views.py [hunk 1]
```diff
@@ -1,20 +1,29 @@
...
+    UserDataReadSerializer
+)
 
 User = get_user_model()
 
 
 class RegistrationViewSet(mixins.CreateModelMixin, viewsets.GenericViewSet):
     queryset = User.objects.all()
-    permission_classes = (AllowAny,)
+    permission_classes = [AllowAny]
```
Comment (jacoor) [inline Dshop/apps/users/api_views.py]: Please revert this change. Tuples, as not editable, are safer in the long run.

File: Dshop/apps/users/api_views.py [hunk 2]
```diff
@@ -25,7 +34,7 @@ def perform_create(self, serializer):
 
 
 class LoginView(GenericAPIView):
-    permission_classes = (AllowAny,)
+    permission_classes = [AllowAny]
```
Comment (jacoor) [inline Dshop/apps/users/api_views.py]: Please revert this change. Tuples, as not editable, are safer in the long run.

File: Dshop/apps/users/api_views.py [hunk 3]
```diff
@@ -45,4 +54,53 @@ def post(self, request):
             return Response(data, status=status.HTTP_200_OK)
         else:
             msg = {'detail': _('Unable to retrieve user auth token')}
-            return Response(msg, status=status.HTTP_400_BAD_REQUEST)
\ No newline at end of file
+            return Response(msg, status=status.HTTP_400_BAD_REQUEST)
+
+
+class LogoutView(GenericAPIView):
+    permission_classes = [AllowAny]
```
Comment (jacoor) [inline Dshop/apps/users/api_views.py]: Tuple, please

File: Dshop/apps/users/api_views.py [hunk 4]
```diff
@@ -45,4 +54,53 @@ def post(self, request):
...
\ No newline at end of file
+            return Response(msg, status=status.HTTP_400_BAD_REQUEST)
+
+
+class LogoutView(GenericAPIView):
+    permission_classes = [AllowAny]
+    serializer_class = EmptySerializer
+    throttle_scope = 'logout'
+
+    def post(self, request):
```
Comment (jacoor) [inline Dshop/apps/users/api_views.py]: I would suggest using GET method here, not POST. You are not sending any data after all, right?

File: Dshop/apps/users/api_views.py [hunk 5]
```diff
@@ -45,4 +54,53 @@ def post(self, request):
...
+
+        return Response(
+            {'detail': _('Successfully logged out.')},
+            status=status.HTTP_200_OK,
+        )
+
+
+class PasswordChangeView(GenericAPIView):
+    serializer_class = PasswordChangeSerializer
+    permission_classes = [IsAuthenticated]
```
Comment (jacoor) [inline Dshop/apps/users/api_views.py]: Tuple please

File: Dshop/apps/users/api_views.py [hunk 6]
```diff
@@ -45,4 +54,53 @@ def post(self, request):
...
+    def post(self, request):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        request.user.set_password(serializer.validated_data['new_password'])
+        request.user.save()
+        return Response({'detail': _('New password has been saved.')}, status=status.HTTP_200_OK)
+
+
+class UserDataChangeView(mixins.RetrieveModelMixin, mixins.UpdateModelMixin, viewsets.GenericViewSet):
+    permission_classes = [IsAuthenticated]
```
Comment (jacoor) [inline Dshop/apps/users/api_views.py]: Tuple, please.

File: Dshop/apps/users/models.py [hunk 7]
```diff
@@ -1,28 +1,30 @@
...
         return self.name
 
 
 class CustomUser(models.Model):
-    user = models.OneToOneField(User, on_delete=models.CASCADE)
-    first_name = models.CharField(max_length=100)
-    last_name = models.CharField(max_length=100)
-    address = models.CharField(max_length=150)
+    user = models.OneToOneField(User, on_delete=models.CASCADE, related_name='customusers', verbose_name=_('User'))
+    first_name = models.CharField(max_length=100, blank=True, verbose_name=_('First name'))
```
Comment (jacoor) [inline Dshop/apps/users/models.py]: Why blank=True? It is required. This goes for all the fields. 
This task was to add API to manage users, not to change business logic around user models. 
Please remove all the blank=True introduced.

File: Dshop/apps/users/serializers.py [hunk 8]
```diff
@@ -1,11 +1,17 @@
...
 from django.utils.translation import gettext_lazy as _
 from rest_framework import serializers
 from rest_framework.serializers import ValidationError
 
+from .models import CustomUser
+
 User = get_user_model()
 
 
+class EmptySerializer(serializers.Serializer):
```
Comment (jacoor) [inline Dshop/apps/users/serializers.py]: Is it used anywhere? If no, please remove.

File: Dshop/apps/users/serializers.py [hunk 9]
```diff
@@ -49,3 +56,53 @@ def validate(self, attrs):
...
+            password_validation.validate_password(attrs['new_password'])
+        return attrs
+
+
+class UserDataReadSerializer(serializers.ModelSerializer):
+    email = serializers.EmailField(source='user.email')
+
+    class Meta:
+        model = CustomUser
+        fields = '__all__'
```
Comment (jacoor) [inline Dshop/apps/users/serializers.py]: Please, never use __all__
Please change it to list of fields.
Using this is dangerous - example - we will introduce PESEL or other personal information. And it ill automatically show in all the places. 
Please replace it with list of fields.

File: Dshop/apps/users/serializers.py [hunk 10]
```diff
@@ -49,3 +56,53 @@ def validate(self, attrs):
...
+
+
+class UserDataChangeSerializer(UserDataReadSerializer):
+    email = serializers.EmailField(required=False)
+
+    def update(self, instance, validated_data):
+        # Update CustomUser fields
```
Comment (jacoor) [inline Dshop/apps/users/serializers.py]: You can try to go with accepted answer from here
https://www.appsloveworld.com/django/100/144/django-update-fields-inside-django-model-object-using-kwargs?expand_article=1
instead of this bunch of copy paste of validated_datea.get...
Or simply use a loop over a list of fields. 
Only update if you want, the code is OK, just a bit of spaghetti.

File: Dshop/apps/users/tests/test_api_logout.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+User = get_user_model()
+
+
+@pytest.fixture
+def logout_url():
+    return reverse('api-logout')
+
+
+@pytest.fixture
+def login_data():
```
Comment (jacoor) [inline Dshop/apps/users/tests/test_api_logout.py]: repeats in other test, please remove or move to conftests.py

File: Dshop/apps/users/tests/test_api_password_change.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+User = get_user_model()
+
+
+@pytest.fixture
+def password_change_url():
+    return reverse('api-password-change')
+
+
+@pytest.fixture
```
Comment (jacoor) [inline Dshop/apps/users/tests/test_api_password_change.py:16]: repeats in other test, please remove or move to conftests.py 
Please also check other fixtures and move to conftest.py if repeated.

File: Dshop/apps/users/tests/test_api_password_change.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    return User.objects.create_user(**login_data)
+
+
+@pytest.fixture
+def user_instance_token(user_instance):
+    return Token.objects.get_or_create(user=user_instance)[0]
+
+
+@pytest.mark.django_db
+def test_password_change_success(api_client, password_change_url, user_instance_token, password_change_data):
```
Comment (jacoor) [inline Dshop/apps/users/tests/test_api_password_change.py]: Please make it two separate tests
1. success
- Would be good to show that I can log in with new password. 
3.  failure.

**Review by jacoor — CHANGES_REQUESTED** [review]

File: Dshop/apps/users/api_views.py (hunk 4 above)
Comment (swietlikm) [reply Dshop/apps/users/api_views.py]: https://www.reddit.com/r/learnprogramming/comments/go5l97/should_logout_in_websites_be_done_using_post/
> Best practice is to use POST with a CSRF token for logout. Otherwise you are leaving a potential path for an attacker to force logout

Comment (jacoor) [reply Dshop/apps/users/api_views.py]: In API? We don't use CSRF token in API calls. First you need also to get user token so... 
It sounds like a good practice in plain HTML pages, not API scopes.  That is also what is this reddit post talkingabout - browser: "So if a webpage has a /logout link on a logout button or somewhere, the browser might log out the user by visiting that URL beforehand by trying to cache data."

Also, how would you get user token first to force logout? 
This is weird... if you get user token to bypass auth there is no point to log out since you already can impostor the user./

File: Dshop/apps/users/serializers.py (hunk 8 above)
Comment (swietlikm) [reply Dshop/apps/users/serializers.py]: @jacoor 
if i don't use serializer in LogoutView then i receive this warning: 

`D:\Py\dshop\Dshop\Dshop\apps\users\api_views.py: Error [LogoutView]: exception raised while getting serializer. Hint: Is get_serializer_class() returning None or is get_queryset() not working without a request? Ignoring the view for now. (Exception: 'LogoutView' sho
uld either include a `serializer_class` attribute, or override the `get_serializer_class()` method.)
`

I found out [here](https://gutsytechster.wordpress.com/2019/11/12/user-auth-operations-in-drf-login-logout-register-change-password/) that they were using EmptySerializer to handle LogoutView.
What do you think?

File: Dshop/apps/users/models.py (hunk 7 above)
Comment (swietlikm) [reply Dshop/apps/users/models.py]: Ok, i will delete blank=true, i would like to remove date_joined as it is already included in standard User model

File: Dshop/apps/users/serializers.py (hunk 9 above)
Comment (swietlikm) [reply Dshop/apps/users/serializers.py]: Good advice, thank you

**Review by jacoor — DISMISSED** [review]

**Review by jacoor — APPROVED** [review]
